src/controller/CommissioningWindowOpener.h

```diff
@@ -0,0 +1,172 @@
+#include <setup_payload/SetupPayload.h>
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace Controller {
+
+// Passing SetupPayload by value on purpose, in case a consumer decides to reuse
+// this object from inside the callback.
+typedef void (*OnOpenCommissioningWindow)(void * context, NodeId deviceId, CHIP_ERROR status, SetupPayload payload);
+typedef void (*OnOpenBasicCommissioningWindow)(void * context, NodeId deviceId, CHIP_ERROR status);
+
+/**
+ * A helper class to open a commissioning window given some parameters.
+ */
+class CommissioningWindowOpener
+{
+public:
+    CommissioningWindowOpener(DeviceController * controller) :
+        mController(controller), mDeviceConnected(&OnDeviceConnectedCallback, this),
+        mDeviceConnectionFailure(&OnDeviceConnectionFailureCallback, this)
+    {}
+
+    enum class CommissioningWindowOption : uint8_t
+    {
+        kOriginalSetupCode = 0,
+        kTokenWithRandomPIN,
+        kTokenWithProvidedPIN,
+    };
+
+    /*
+     * @brief
+     *   Try to look up the device attached to our controller with the given
+     *   node id and ask it to re-enter commissioning mode with its original
+     *   PASE verifier, discriminator, etc. The device will exit commissioning
+     *   mode after a successful commissioning, or after the given `timeout`
+     *   time.
+     *
+     * @param[in] deviceId The device Id.
+     * @param[in] timeout  The commissioning mode should terminate after this much time.
+     * @param[in] callback The callback to call once the commissioning window is
+     *                     open or if an error occurs.
+     */
+    CHIP_ERROR OpenBasicCommissioningWindow(NodeId deviceId, System::Clock::Seconds16 timeout,
+                                            Callback::Callback<OnOpenBasicCommissioningWindow> * callback);
+
+    /**
+     * @brief
+     *   Try to look up the device attached to our controller with the given
+     *   node id and ask it to re-enter commissioning mode with a PASE verifier
+     *   derived from the given information and the given discriminator. The
+     *   device will exit commissioning mode after a successful commissioning,
+     *   or after the given `timeout` time.
+     *
+     * @param[in] deviceId      The device Id.
+     * @param[in] timeout       The commissioning mode should terminate after this much time.
+     * @param[in] iteration     The PAKE iteration count associated with the PAKE Passcode ID and ephemeral
+     *                          PAKE passcode verifier to be used for this commissioning.
+     * @param[in] discriminator The long discriminator for the DNS-SD advertisement.
+     * @param[in] setupPIN      The setup PIN to use, or NullOptional to use a randomly-generated one.
+     * @param[in] callback      The function to be called on success or failure of opening of commissioning window.
+     * @param[out] payload      The setup payload, not including the VID/PID bits, that is generated based on the
```

Comment:
This comment seems to conflict with `readVIDPIDAttributes` description

Reply:
In what way?

There are two setup payloads involved.  One is handed out synchronously from this call.  One is delivered async via the callback.  The former can't have a VID/PID in it, because it needs to be sync.  The latter will have them if the caller asked for them.

Reply:
Updated the docs to make this clearer.